staging/src/k8s.io/legacy-cloud-providers/azure/clients/armclient/azure_armclient.go

```diff
@@ -0,0 +1,543 @@
+// +build !providerless
+
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package armclient
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+	"unicode"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure"
+
+	"k8s.io/client-go/pkg/version"
+	"k8s.io/klog"
+	"k8s.io/legacy-cloud-providers/azure/retry"
+)
+
+var _ Interface = &Client{}
+
+// Client implements ARM client Interface.
+type Client struct {
+	client  autorest.Client
+	backoff *retry.Backoff
+
+	baseURI      string
+	apiVersion   string
+	clientRegion string
+}
+
+// New creates a ARM client
+func New(authorizer autorest.Authorizer, baseURI, userAgent, apiVersion, clientRegion string, clientBackoff *retry.Backoff) *Client {
+	restClient := autorest.NewClientWithUserAgent(userAgent)
+	restClient.PollingDelay = 5 * time.Second
+	restClient.RetryAttempts = 3
+	restClient.RetryDuration = time.Second * 1
+	restClient.Authorizer = authorizer
+
+	if userAgent == "" {
+		restClient.UserAgent = GetUserAgent(restClient)
+	}
+
+	backoff := clientBackoff
+	if backoff == nil {
+		// 1 steps means no retry.
+		backoff = &retry.Backoff{
+			Steps: 1,
+		}
+	}
+
+	return &Client{
+		client:       restClient,
+		baseURI:      baseURI,
+		backoff:      backoff,
+		apiVersion:   apiVersion,
+		clientRegion: NormalizeAzureRegion(clientRegion),
+	}
+}
+
+// GetUserAgent gets the autorest client with a user agent that
+// includes "kubernetes" and the full kubernetes git version string
+// example:
+// Azure-SDK-for-Go/7.0.1 arm-network/2016-09-01; kubernetes-cloudprovider/v1.17.0;
+func GetUserAgent(client autorest.Client) string {
+	k8sVersion := version.Get().GitVersion
+	return fmt.Sprintf("%s; kubernetes-cloudprovider/%s", client.UserAgent, k8sVersion)
+}
+
+// NormalizeAzureRegion returns a normalized Azure region with white spaces removed and converted to lower case
+func NormalizeAzureRegion(name string) string {
```

Comment:
ut

Reply:
Would add a unit test.

Reply:
Added the unit tests in the new pushes. PTAL